costmap_2d/plugins/shadow_layer.cpp

```diff
@@ -0,0 +1,358 @@
+{
+  ROS_DEBUG("Updating shadow layer costs");
+  if (!enabled_) {
+    return;
+  }
+
+  calculateShadows(master_grid);
+}
+
+void ShadowLayer::calculateShadows(costmap_2d::Costmap2D& master_grid)
+{
+  ROS_DEBUG("Calculating shadows");
+  // First, reset the costmap.
+  resetMaps();
+
+  // Raytrace the observations into the map
+  raytraceShadowObservation(master_grid);
+
+  // Calculate all of the spots where there could be a shadow.
+  calculateShadowedObjects();
+
+  if (publish_shadow_objects_)
+  {
+    publishShadowObjects();
+  }
+}
+
+void ShadowLayer::publishShadowObjects()
+{
+  if (!shadowed_objects_)
+  {
+    return;
+  }
+
+  ROS_DEBUG("Publishing display objects");
+  visualization_msgs::Marker sphere_list;
+  sphere_list.header.frame_id = global_frame_;
+  sphere_list.header.stamp = ros::Time::now();
+  sphere_list.ns = "shadows/" + global_frame_;
+  sphere_list.action= visualization_msgs::Marker::ADD;
+  sphere_list.pose.orientation.w= 1.0;
+
+  sphere_list.id = 0;
+
+  sphere_list.type = visualization_msgs::Marker::SPHERE_LIST;
+
+  sphere_list.scale.x = min_shadow_size_;
+  sphere_list.scale.y = min_shadow_size_;
+  sphere_list.scale.z = min_shadow_size_;
+
+  sphere_list.color.r = 1.0f;
+  sphere_list.color.g = 1.0f;  
+  sphere_list.color.a = 1.0;
+
+  for (auto pt : (*shadowed_objects_))
+  {
+    sphere_list.points.push_back(pt);
+  }
+  visualization_publisher_.publish(sphere_list);
+}
+
+void ShadowLayer::raytraceShadowObservation(costmap_2d::Costmap2D& master_grid)
+{
+  if (!shadowed_points_)
+  {
+    shadowed_points_ = std::make_shared<std::vector<unsigned int>>();
+  }
+  // Clear out the old shadow points
+  shadowed_points_->clear();
+
+  // Check if the observation is valid.
+  if (!shadow_observation_)
+  {
+    ROS_WARN_THROTTLE(1.0, "Cannot raytrace in shadow layer because there is no shadow observation.");
+    return;
+  }
+
+  // 
+  double ox = shadow_observation_->origin_.x;
+  double oy = shadow_observation_->origin_.y;
+  pcl::PointCloud < pcl::PointXYZ > cloud = *(shadow_observation_->cloud_);
+
+  // get the map coordinates of the origin of the sensor
+  unsigned int x0, y0;
+  if (!worldToMap(ox, oy, x0, y0))
+  {
+    ROS_WARN_THROTTLE(
+        1.0, "The origin for the sensor at (%.2f, %.2f) is out of map bounds. So, the costmap cannot raytrace for it.",
+        ox, oy);
+    return;
+  }
+
+  // we can pre-compute the enpoints of the map outside of the inner loop... we'll need these later
+  double origin_x = origin_x_, origin_y = origin_y_;
+  double map_end_x = origin_x + size_x_ * resolution_;
+  double map_end_y = origin_y + size_y_ * resolution_;
+
+
+  // for each point in the cloud, we want to trace a line from the origin and clear obstacles along it
+  for (unsigned int i = 0; i < cloud.points.size(); ++i)
+  {
+    double wx = cloud.points[i].x;
+    double wy = cloud.points[i].y;
+
+    double a = wx - ox;
+    double b = wy - oy;
+    double ray_length = sqrt(a * a + b * b);
```

Comment:
Is this used anywhere (if it is, hypot may be cheaper)?

Reply:
Good catch - completely unused and wasting cycles.